### PR TITLE
Obj Construction System

### PR DESCRIPTION
**Pull request by Cyberboss**
Gonna give this another go without refactoring everything at once

### CODE REVIEW

**Comment by praisenarsie** [comment]
hooray!

**Comment by Cyberboss** [comment]
Here's a test case. Hopefully able to catch all cases in our codebase:

Lets say we have some buildable item:

Start from 2 wood, begin with the crafting menu, creates an item with a max_integ of 100

then it requires 2 more wood (at this point a screwdriver can be used to repair it). It can be attack_hand'd to deconstruct it. At this point it's max_integ is 200

then it requires a diamond. (optionally it can accept a stunbaton(only during this step) and a taser which will give it different features (attack_handing it at this point should give you the option to remove the taser/baton/deconstruct to the next step (removing items like this should always give back the exact item that was put in))). After this step it is no longer an /item and falls to the users drop_location(). After this step in can no longer be screwdriver repaired.

Then it requires both two metal and two glass, which can be applied in any order. A screwdriver will deconstruct it from this phase (Only after any intermediary metal/glass is removed by attack_handing it). After this step the taser can no longer be applied.

Then a wrench is required. or attack_hand can remove the metal/glass. A welding tool will repair the object during and after this phase.

Then it requires a power cell which can be of any type and affects the final device's power level. Wrenching from this phase will deconstruct it.

Finally a screwdriver will finish the device, only at this point is it operational and can consume power, interact with shit, etc.

If the final device has a taser it will stun mobs exactly 2 tiles away

If the final device has a stunbaton it will stun mobs next to it

each phase/sidegrade affects the appearance differently. At any point examines should be able to state construction/deconstuction steps, what's required/what's optional/what it has. Deconstruction should always spit out the exact items you put into it (and the power_cell should have the appropriate power based on how much the device used). (or just the right type if it was a stack material)

**Comment by Bawhoppen** [comment]
I'm glad you're doing this for us all

**Comment by MrStonedOne** [comment]
How does that fit in with slap crafting @Cyberboss

**Comment by optimumtact** [comment]
he's literally just described slap crafting you muppet

**Comment by Cyberboss** [comment]
currently coding my described scenario in currentcode

**Comment by Iamgoofball** [comment]
oh

@MrStonedOne this is devicecrafting not other shit

File: code/game/objects/obj_construction_test.dm [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	icon_state = "box"
+	item_state = "syringe_kit"
+	lefthand_file = 'icons/mob/inhands/equipment/medical_lefthand.dmi'
+	righthand_file = 'icons/mob/inhands/equipment/medical_righthand.dmi'
+	resistance_flags = FLAMMABLE
+	max_integrity = 100
+	var/construction_phase = 1
+	var/is_being_constructed
+	var/obj/item/melee/baton/baton
+	var/obj/item/gun/energy/taser/taser
```
Comment (Iamgoofball) [inline code/game/objects/obj_construction_test.dm]: wait nvm this is hardcoded garbage

Comment (Bawhoppen) [reply code/game/objects/obj_construction_test.dm]: it's the test file goof

**Comment by duncathan** [comment]
@Cyberboss

**Comment by optimumtact** [comment]
This isn't trivial, there's so many methods of construction that it will take quite a bit of careful thought to come up with a system that can support everything.

@Cyberboss to get this into a working state faster, maybe just try to support some basic recipes only first

As we port things to the system we can extend it on the fly to support the more complex systems, as long as it's a finite state machine we should be able finangle everything we need in eventually.

**Comment by Cyberboss** [comment]
I'll get back to dis on Monday

**Comment by Bawhoppen** [comment]
Will this fix #12267?

**Comment by Cyberboss** [comment]
s o o n (tm)

**Comment by Militaires** [comment]
# i proposed a new catch-all for cyberboss and he asked me to explain its specifics right here, so here goes:

every component ingame falls into a series of broad categories such as [CHASSIS], [HULL], [COMPONENT] or [ELECTRONICS]

**OPTION 1 for circuits:** circuits/[ELECTRONICS] are subdivided even further in order to make the system even more modular, with each device having a set of possible circuits each either changing one or more variables or giving it new unprecedented functionality, so let's say in the [MECH] category you can find [MECH:MOVEMENT] circuits, [MECH:COMBAT] circuits, and a high-tech [MECH:BLUESPACE] circuit which allows the bluespace artillery [COMPONENT] installed on the mecha to fire.

**OPTION 2 for circuits:** circuits arent specific at all and are just like manipulators; used for a wide variety of purposes but never deviating to give a unique function or ability, as such you'd use the same circuit in a mecha as the one in the blender with the very general name [ELECTRONICS BOARD], this sacrifices customization for simplicity.

case 1: (mech)

you can choose a [CHASSIS] made out of say, metal, plasteel, bananium or glass if you're a madman, then you decide to add another [CHASSIS] in order to reinforce it, this time made out of uranium.

you print two general [ELECTRONIC] circuits both designed for [MECH: MOVEMENT] which makes the mecha very responsive to your buttons but it takes up space in the CPU so you can only put one [MECH: LIFE_SUPPORT] circuit and no [MECH:COMBAT] circuit

you create four simple steel limbs, each one of them a part of the [COMPONENTS] category, and attach them. (read the last part of this comment if you want to expand on this)

you make and attach one diamond [HULL] then plate that with a plasteel [HULL] to finish it off, this results in a very responsive, medium-speed mecha that is ill-suited for combat.

case 2: blender

you make a steel blender [CHASSIS] for that nice blendtec sheen, install [ELECTRONICS:BLENDER_CONTROL] boards, you go get diamonds from mining and turn them into diamond blender blades, which is a [COMPONENT], then you top it all off with a plasma-glass [HULL] so that it's visible.

case 3: space pod

you make a bananium [CHASSIS], add [POD:MOVEMENT] circuits ... 

## EXTREME AUTISM, CYBERBOSS BEWARE:
what IF you even make the system more modular, more autistic?

if you're a willing enough madman, you might want to make every [COMPONENT] its own device of its own right, that means for example, space pod thrusters would have their own [CHASSIS], [ELECTRONICS], and [HULL].

the mecha limbs i mentioned earlier would use their own electronics, their own hull, and perhaps, if you're insane, their own component (a component within a component) consisting of say thrusters at the bottom of the feet to make the mech "jump"

**Comment by Bawhoppen** [comment]
Please don't let this die

**Comment by Cyberboss** [comment]
It's on my list, behind like 20 other things

**Comment by Bawhoppen** [comment]
I feel like it died.

**Pull request closed without merging**